File: providers/aws/src/main/java/cloudspec/aws/ec2/resource/nested/EC2IpPermission.java
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 * #L%
 */
package cloudspec.aws.ec2.resource.nested;

import cloudspec.annotation.PropertyDefinition;
import software.amazon.awssdk.services.ec2.model.IpPermission;
import software.amazon.awssdk.services.ec2.model.PrefixListId;

import java.util.Collections;
import java.util.List;
import java.util.Objects;
import java.util.Optional;
import java.util.stream.Collectors;

public class EC2IpPermission {
    @PropertyDefinition(
            name = "from_port",
            description = "The start of port range for the TCP and UDP protocols, or an ICMP/ICMPv6 type number"
    )
    private final Integer fromPort;

    @PropertyDefinition(
            name = "ip_protocol",
            description = "The IP protocol name or number"
    )
    private final String ipProtocol;

    @PropertyDefinition(
            name = "ip_ranges",
            description = "The IPv4 ranges"
    )
    private final List<EC2IpRange> ipRanges;

    @PropertyDefinition(
            name = "ipv6_ranges",
            description = "[VPC only] The IPv6 ranges"
    )
    private final List<EC2Ipv6Range> ipv6Ranges;

    @PropertyDefinition(
            name = "prefix_list_ids",
            description = "[VPC only] The prefix list IDs for an AWS service"
    )
    private final List<String> prefixListIds;

    @PropertyDefinition(
            name = "to_port",
            description = "The end of port range for the TCP and UDP protocols, or an ICMP/ICMPv6 code"
    )
    private final Integer toPort;

    @PropertyDefinition(
            name = "user_groups",
            description = "The security group and AWS account ID pairs"
    )
    private final List<EC2UserIdGroupPair> userIdGroupPairs;

    public EC2IpPermission(Integer fromPort, String ipProtocol, List<EC2IpRange> ipRanges, List<EC2Ipv6Range> ipv6Ranges,
                           List<String> prefixListIds, Integer toPort, List<EC2UserIdGroupPair> userIdGroupPairs) {
        this.fromPort = fromPort;
        this.ipProtocol = ipProtocol;
        this.ipRanges = ipRanges;
        this.ipv6Ranges = ipv6Ranges;
        this.prefixListIds = prefixListIds;
        this.toPort = toPort;
        this.userIdGroupPairs = userIdGroupPairs;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        EC2IpPermission that = (EC2IpPermission) o;
        return Objects.equals(fromPort, that.fromPort) &&
                Objects.equals(ipProtocol, that.ipProtocol) &&
                Objects.equals(ipRanges, that.ipRanges) &&
                Objects.equals(ipv6Ranges, that.ipv6Ranges) &&
                Objects.equals(prefixListIds, that.prefixListIds) &&
                Objects.equals(toPort, that.toPort) &&
                Objects.equals(userIdGroupPairs, that.userIdGroupPairs);
    }

    @Override
    public int hashCode() {
        return Objects.hash(fromPort, ipProtocol, ipRanges, ipv6Ranges, prefixListIds, toPort, userIdGroupPairs);
    }

    public static List<EC2IpPermission> fromSdk(List<IpPermission> ipPermissions) {
        return Optional.ofNullable(ipPermissions)
                .orElse(Collections.emptyList())
                .stream()
                .map(EC2IpPermission::fromSdk)
                .collect(Collectors.toList());
    }

    public static EC2IpPermission fromSdk(IpPermission ipPermission) {
        return Optional.ofNullable(ipPermission)
                .map(v -> new EC2IpPermission(
                                v.fromPort(),
                                v.ipProtocol(),
                                EC2IpRange.fromSdk(v.ipRanges()),
                                EC2Ipv6Range.fromSdk(v.ipv6Ranges()),
                                prefixListIdsFromSdk(v.prefixListIds()),
                                v.toPort(),
                                EC2UserIdGroupPair.fromSdk(v.userIdGroupPairs())
                        )
                )
                .orElse(null);
    }

    public static List<String> prefixListIdsFromSdk(List<PrefixListId> prefixListIds) {
        return Optional.ofNullable(prefixListIds)
                .orElse(Collections.emptyList())
                .stream()
                .map(PrefixListId::prefixListId)
                .collect(Collectors.toList());
    }
}

File: providers/aws/src/main/kotlin/cloudspec/aws/ec2/nested/EC2IpPermission.kt
/*-
 * #%L
 * CloudSpec AWS Provider
 * %%
 * Copyright (C) 2020 Ezequiel Foncubierta
 * %%
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * #L%
 */
package cloudspec.aws.ec2.nested

import cloudspec.annotation.PropertyDefinition

data class EC2IpPermission(
        @property:PropertyDefinition(
                name = PROP_FROM_PORT,
                description = PROP_FROM_PORT_D
        )
        val fromPort: Int?,

        @property:PropertyDefinition(
                name = PROP_IP_PROTOCOL,
                description = PROP_IP_PROTOCOL_D
        )
        val ipProtocol: String?,

        @property:PropertyDefinition(
                name = PROP_IP_RANGES,
                description = PROP_IP_RANGES_D
        )
        val ipRanges: List<EC2IpRange>?,

        @property:PropertyDefinition(
                name = PROP_IPV6_RANGES,
                description = PROP_IPV6_RANGES_D
        )
        val ipv6Ranges: List<EC2Ipv6Range>?
        ,
        @property:PropertyDefinition(
                name = PROP_PREFIX_LIST_IDS,
                description = PROP_PREFIX_LIST_IDS_D
        )
        val prefixListIds: List<String>?,

        @property:PropertyDefinition(
                name = PROP_TO_PORT,
                description = PROP_TO_PORT_D
        )
        val toPort: Int?,

        @property:PropertyDefinition(
                name = PROP_USER_GROUPS,
                description = PROP_USER_GROUPS_D
        )
        val userIdGroupPairs: List<EC2UserIdGroupPair>?
) {
    companion object {
        const val PROP_FROM_PORT = "from_port"
        const val PROP_FROM_PORT_D = "The start of port range for the TCP and UDP protocols, or an ICMP/ICMPv6 type number"
        const val PROP_IP_PROTOCOL = "ip_protocol"
        const val PROP_IP_PROTOCOL_D = "The IP protocol name or number"
        const val PROP_IP_RANGES = "ip_ranges"
        const val PROP_IP_RANGES_D = "The IPv4 ranges"
        const val PROP_IPV6_RANGES = "ipv6_ranges"
        const val PROP_IPV6_RANGES_D = "[VPC only] The IPv6 ranges"
        const val PROP_PREFIX_LIST_IDS = "prefix_list_ids"
        const val PROP_PREFIX_LIST_IDS_D = "[VPC only] The prefix list IDs for an AWS service"
        const val PROP_TO_PORT = "to_port"
        const val PROP_TO_PORT_D = "The end of port range for the TCP and UDP protocols, or an ICMP/ICMPv6 code"
        const val PROP_USER_GROUPS = "user_groups"
        const val PROP_USER_GROUPS_D = "The security group and AWS account ID pairs"
    }
}
